near-sdk/src/store/index_map.rs

```diff
@@ -0,0 +1,152 @@
+
+    /// Sets a value at a given index to the value provided. If none is provided, this index will
+    /// be removed from storage.
+    pub fn set(&mut self, index: u32, value: Option<T>) {
+        let entry = self.cache.get_mut(index);
+        match entry.get_mut() {
+            Some(entry) => *entry.value_mut() = value,
+            None => {
+                let _ = entry.set(CacheEntry::new_modified(value));
+            }
+        }
+    }
+}
+
+impl<T> IndexMap<T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    fn deserialize_element(raw_element: &[u8]) -> T {
+        T::try_from_slice(raw_element)
+            .unwrap_or_else(|_| env::panic_str(ERR_ELEMENT_DESERIALIZATION))
+    }
+
+    /// Returns the element by index or `None` if it is not present.
+    pub fn get(&self, index: u32) -> Option<&T> {
+        let entry = self.cache.get(index).get_or_init(|| {
+            let mut buf = Vec::with_capacity(self.prefix.len() + 4);
+            Self::index_to_lookup_key(&self.prefix, index, &mut buf);
+            let storage_bytes = env::storage_read(&buf);
+            let value = storage_bytes.as_deref().map(Self::deserialize_element);
+            CacheEntry::new_cached(value)
+        });
+        entry.value().as_ref()
+    }
+
+    /// Returns a mutable reference to the element at the `index` provided.
+    pub(crate) fn get_mut_inner(&mut self, index: u32) -> &mut CacheEntry<T> {
+        let prefix = &self.prefix;
+        let entry = self.cache.get_mut(index);
+        entry.get_or_init(|| {
+            let mut key = Vec::with_capacity(prefix.len() + 4);
+            Self::index_to_lookup_key(prefix, index, &mut key);
+            let storage_bytes = env::storage_read(&key);
+            let value = storage_bytes.as_deref().map(Self::deserialize_element);
+            CacheEntry::new_cached(value)
+        });
+        let entry = entry.get_mut().unwrap();
+        entry
+    }
+
+    /// Returns a mutable reference to the element at the `index` provided.
+    pub fn get_mut(&mut self, index: u32) -> Option<&mut T> {
+        let entry = self.get_mut_inner(index);
+        entry.value_mut().as_mut()
+    }
+
+    pub fn swap(&mut self, a: u32, b: u32) {
+        if a == b {
+            // Short circuit if indices are the same, also guarantees uniqueness below
+            return;
+        }
+
+        let val_a = self.get_mut_inner(a).replace(None);
+        let val_b = self.get_mut_inner(b).replace(val_a);
+        self.get_mut_inner(a).replace(val_b);
+    }
+
+    /// Inserts a element at `index`, returns the evicted element.
+    pub fn replace(&mut self, index: u32, element: T) -> Option<T> {
+        self.get_mut_inner(index).replace(Some(element))
+    }
+}
+
+impl<T> fmt::Debug for IndexMap<T>
+where
+    T: BorshSerialize + BorshDeserialize + fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("IndexMap").field("prefix", &self.prefix).finish()
+    }
+}
```

Comment:
Probably best to have some tests for `IndexMap`, even if `Vector` already covers it at a higher level, just so we can see some of the basic features it can cover

Reply:
yeah, you're definitely right I should, I was just being lazy because it's an internal type. Adding now